src/main/scala/io/iohk/ethereum/forkid/ForkIdValidator.scala

```diff
@@ -0,0 +1,129 @@
+package io.iohk.ethereum.forkid
+
+import akka.util.ByteString
+import cats.Monad
+import cats.data.EitherT._
+import cats.implicits._
+import io.iohk.ethereum.utils.BigIntExtensionMethods._
+import io.iohk.ethereum.utils.BlockchainConfig
+import io.iohk.ethereum.utils.ByteUtils._
+import monix.eval.Task
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+import java.util.zip.CRC32
+
+sealed trait ForkIdValidationResult
+case object Connect extends ForkIdValidationResult
+case object ErrRemoteStale extends ForkIdValidationResult
+case object ErrLocalIncompatibleOrStale extends ForkIdValidationResult
+
+object ForkIdValidator {
+
+  implicit val unsafeLogger = Slf4jLogger.getLogger[Task]
+
+  val maxUInt64 = (BigInt(0x7fffffffffffffffL) << 1) + 1 // scalastyle:ignore magic.number
+
+  def validatePeer[F[_]: Monad: Logger](
+      genesisHash: ByteString,
+      config: BlockchainConfig
+  )(currentHeight: BigInt, remoteForkId: ForkId): F[ForkIdValidationResult] = {
+    val forks = ForkId.gatherForks(config)
+    validatePeer[F](genesisHash, forks)(currentHeight, remoteForkId)
+  }
+
+  def validatePeer[F[_]: Monad: Logger](
+      genesisHash: ByteString,
+      forks: List[BigInt]
+  )(currentHeight: BigInt, remoteId: ForkId): F[ForkIdValidationResult] = {
+    val checksums: Vector[BigInt] = calculateCheckchecksums(genesisHash, forks)
+
+    // find the first unpassed fork and it's index
+    val (unpassedFork, i) = (forks :+ maxUInt64).zipWithIndex.find { case (fork, _) => currentHeight < fork }.get
+
+    // The checks are left biased -> whenever a result is found we need to short circuit
+    val validate: F[Option[ForkIdValidationResult]] = (for {
```

Comment:
This method looks a lot more complicated than it is. How about moving the tracing into the `check` methods, for a start? (Do we still need the tracing btw?)

Reply:
> This method looks a lot more complicated than it is.

This is vague, can you elaborate on that? 

>  How about moving the tracing into the check methods, for a start?

It could be done, but I can't see how moving logging instructions around is going to make the method less complicated.

> (Do we still need the tracing btw?)

Logging is part of our DoD and it might get useful when testing on testnet, so I wouldn't remove it.